Lib/cu2qu/ufo.py

```diff
@@ -0,0 +1,198 @@
+# Copyright 2015 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+"""Converts cubic bezier curves to quadratic splines.
+
+Conversion is performed such that the quadratic splines keep the same end-curve
+tangents as the original cubics. The approach is iterative, increasing the
+number of segments for a spline until the error gets below a bound.
+
+Respective curves from multiple fonts will be converted at once to ensure that
+the resulting splines are interpolation-compatible.
+"""
+
+
+from __future__ import print_function, division, absolute_import
+
+from fontTools.pens.basePen import AbstractPen
+from cu2qu import curve_to_quadratic, curves_to_quadratic
+
+__all__ = ['fonts_to_quadratic', 'font_to_quadratic']
+
+DEFAULT_MAX_ERR = 0.0025
+
+
+_zip = zip
+def zip(*args):
+    """Ensure each argument to zip has the same length."""
+
+    if len(set(len(a) for a in args)) != 1:
+        msg = 'Args to zip in cu2qu should have equal lengths: '
+        raise ValueError(msg + ' '.join(str(a) for a in args))
+    return _zip(*args)
+
+
+class GetSegmentsPen(AbstractPen):
+    """Pen to collect segments into lists of points for conversion.
+
+    Curves always include their initial on-curve point, so some points are
+    duplicated between segments.
+    """
+
+    def __init__(self):
+        self._last_pt = None
+        self.segments = []
+
+    def _add_segment(self, tag, *args):
+        if tag in ['move', 'line', 'qcurve', 'curve']:
+            self._last_pt = args[-1]
+        self.segments.append((tag, args))
+
+    def moveTo(self, pt):
+        self._add_segment('move', pt)
+
+    def lineTo(self, pt):
+        self._add_segment('line', pt)
+
+    def qCurveTo(self, *points):
+        self._add_segment('qcurve', self._last_pt, *points)
+
+    def curveTo(self, *points):
+        self._add_segment('curve', self._last_pt, *points)
+
+    def closePath(self):
+        self._add_segment('close')
+
+    def endPath(self):
+        self._add_segment('end')
+
+    def addComponent(self, glyphName, transformation):
+        self._add_segment('component', glyphName, transformation)
+
+
+def _get_segments(glyph):
+    """Get a glyph's segments as extracted by GetSegmentsPen."""
+
+    pen = GetSegmentsPen()
+    glyph.draw(pen)
+    return pen.segments
+
+
+def _set_segments(glyph, segments):
+    """Draw segments as extracted by GetSegmentsPen back to a glyph."""
+
+    glyph.clearContours()
```

Comment:
There's also another issue here which is somewhat related to this.
When working from within a UFO2 environment (robofab/master or defcon/master), the current `cu2qu.ufo` writes the anchors twice.
This is because in UFO2-based libraries the pens also "draw" anchors (i.e.interpreted as single points with a "move" type and a "name" attribute) as well as contours and components.
https://github.com/typesupply/defcon/blob/ufo3/Lib/defcon/objects/glyph.py#L363-L373
https://github.com/typesupply/defcon/blob/master/Lib/defcon/objects/glyph.py#L304-L313

When using defcon/ufo3, anchors are separate objects whose data is not passed along by the `drawPoints`:
https://github.com/typesupply/defcon/blob/ufo3/Lib/defcon/objects/glyph.py#L363-L373

One fix would be to call `glyph.clearAnchors()` but only when drawing on a UFO2-kind of glyph (but not when the glyph is UFO3, since in this case the pens would not re-draw the anchors).

These are some of the problems that I got stuck into when I was trying to find a general way to use pens to convert glyphs from different environments. I believe that cu2qu, at least in its most general API, should only take care of creating new glyph-like objects (exporting a minimal `draw` and `drawPoints` API) which then the caller will use to draw on the specific "canvas" it is working with.